src/core/shape_inference/include/convolution_shape_inference.hpp

```diff
@@ -444,83 +530,82 @@ bool resolve_auto_pad_for_shape_back_prop(const ConvType* op,
     return status;
 }
 
-
-
-template<class ConvType, class ShapeType>
-void calculate_output_spatial_dims_for_convolution_back_prop_data(
-        const ConvType* op,
-        const ShapeType& input_shape,
-        const ShapeType& filters_shape,
-        const ShapeType& output_rank,
-        ShapeType& output_shape,
-        const int64_t& num_spatial,
-        const Strides& strides,
-        const Strides& dilations,
-        const CoordinateDiff& pads_begin,
-        const CoordinateDiff& pads_end,
-        const CoordinateDiff& output_padding,
-        const int64_t& num_non_spatial_data_dims,
-        const int64_t& num_non_spatial_filter_dims) {
-
-}
-
 template <class T>
 void shape_infer(const ConvolutionBackpropData* op,
                  const CoordinateDiff& pads_begin,
                  const CoordinateDiff& pads_end,
                  const T& output_shape_from_input,
                  const std::vector<T>& input_shapes,
                  std::vector<T>& output_shapes) {
+    size_t num_non_spatial_data_dims = 2, num_non_spatial_filter_dims = 2;
     size_t input_size = input_shapes.size();
     NODE_VALIDATION_CHECK(op, (input_size == 2 || input_size == 3) && output_shapes.size() == 1);
     auto input_shape = input_shapes[0], filters_shape = input_shapes[1];
 
-    const auto& num_spatial = op->m_num_spatial;
-    NODE_VALIDATION_CHECK(op, num_spatial != -1,
+    auto num_spatial = op->m_num_spatial;
+    if (num_spatial == -1 && input_size == 3)
```

Comment:
it looks  like this both case are excludes so use some `if..else` to reduce number of checks and assignments to `num_spatial`

Also this block code (L:545-560) looks same as L:625-642, make some helper to reduce duplicates?

Reply:
this method has varying number of parameters so this is not a duplication